### ISSUE TITLE
Usage of GitHub Artifact Attestations

### ISSUE THREAD

**Issue by pellared** [open-telemetry/sig-security#79]
I propose to check if [GitHub Artifact Attestations](https://github.blog/news-insights/product-news/introducing-artifact-attestations-now-in-public-beta) (which uses Sigstore cosign under the cover) would good for signing different type of release artifacts.

It looks to be the most user-friendly and straightforward approach for signing different type of artifacts (containers and files) when using GitHub. ee [here](https://docs.github.com/en/actions/security-for-github-actions/using-artifact-attestations/using-artifact-attestations-to-establish-provenance-for-builds) how easy it it to use, especially for binaries (or any files). [actions/attest-build-provenance](https://github.com/actions/attest-build-provenance) is a GitHub Action which does the signing. Verification can be done by using [gh CLI](https://cli.github.com/manual/gh_attestation_verify).

I can give it a shot and test it for https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation. However, I would like the Security SIG to make a quick assessment if it is even worth testing.

**Comment by mx-psi** [comment]
We don't use the attestation thing, but we do use cosign on the Collector: https://github.com/open-telemetry/opentelemetry-collector?tab=readme-ov-file#verifying-the-images-signatures

**Comment by pellared** [comment]
> We don't use the attestation thing, but we do use cosign on the Collector: https://github.com/open-telemetry/opentelemetry-collector?tab=readme-ov-file#verifying-the-images-signatures

AFAIK Collector container images are signed but the released executables are not.

**Comment by mx-psi** [comment]
Yes, that's it, we would benefit from using this, just wanted to mention it as a "we are doing *some* of this"

**Comment by pellared** [comment]
@mx-psi, actually I see that you are also publishing `pem` and `sig` files in releases e.g. https://github.com/open-telemetry/opentelemetry-collector-releases/releases/tag/v0.112.0. Therefore, probably it is possibly to verify like described here: https://edu.chainguard.dev/open-source/sigstore/cosign/how-to-verify-file-signatures-with-cosign/

@CodeBlanch, @Kielek told me that you recently added something similar here: https://github.com/open-telemetry/opentelemetry-dotnet/pull/5880. PS. There are no user-facing docs how the users should use it.

At first glance, it looks like using GitHub Artifact Attestations is more user friendly for both maintainers and end-users. 

The other  good part of it is that the generated Sigstore bundle is stored with GitHub and is also written to an immutable transparency log that is publicly readable on the internet as opposed to storing them in GitHub Releases artifacts. This is described here: https://docs.github.com/en/actions/security-for-github-actions/using-artifact-attestations/using-artifact-attestations-to-establish-provenance-for-builds.

**Comment by CodeBlanch** [comment]
@pellared 

> PS. There are no user-facing docs how the users should use it.

I know, sorry. I have an issue tracking that: https://github.com/open-telemetry/opentelemetry-dotnet/issues/5934

It sounds like where .NET is a bit different than Collector is we don't use release/artifacts to store the `pem` and `sig` files. We ship them inside the packages we push to NuGet. Check out: https://nuget.info/packages/OpenTelemetry/1.10.0-rc.1 (browse into the `lib` folder).

So as far as writing them to something immutable and transparent we are kind of already there.

That being said, I'm not opposed to using GitHub Artifact Attestations. Could switch to that or do both.

What I'm hoping is (eventually) NuGet comes around and adds first-class support for SigStore: https://github.com/NuGet/Home/issues/12856.
